copylot/hardware/orca_camera/camera.py

```diff
@@ -77,3 +125,337 @@ def run(self, nb_frame: int = 100000):
+                                      fValue=2)  # fValue = 3 is burst mode, 1 is continuous mode. 2 is start mode.
+            assert abs(v - 2) < 0.00001
+        elif self.master_pulse_mode == 'BURST':
+            v = dcam.prop_setgetvalue(idprop=DCAM_IDPROP.MASTERPULSE_MODE,
+                                      fValue=3)  # fValue = 3 is burst mode, 1 is continuous mode. 2 is start mode.
+            assert abs(v - 3) < 0.00001
+            # set master pulse burst times
+
+            v = dcam.prop_setgetvalue(idprop=DCAM_IDPROP.MASTERPULSE_BURSTTIMES,
+                                      fValue=self.burst_times)
+            assert abs(v - self.burst_times) < 0.00001
+        else:
+            raise ValueError('master pulse mode is set to '+str(self.master_pulse_mode)+';'
+                             'only \'CONTINUOUS\', \'START\' and \'BURST\' are supported.')
+
+        # set master pulse interval
+        v = dcam.prop_setgetvalue(idprop=DCAM_IDPROP.MASTERPULSE_INTERVAL,
+                                  fValue=self.master_pulse_interval)
+        assert abs(v - self.master_pulse_interval) < 0.00001
+
+        # -- set master pulse trigger to be external trigger:
+        if self.master_pulse_trigger == 'EXTERNAL':
+            v = dcam.prop_setgetvalue(idprop=DCAM_IDPROP.MASTERPULSE_TRIGGERSOURCE,
+                                      fValue=1)  # this sets the trigger source to be external.
+            assert abs(v - 1) < 0.00001
+        elif self.master_pulse_trigger == 'SOFTWARE':
+            v = dcam.prop_setgetvalue(idprop=DCAM_IDPROP.MASTERPULSE_TRIGGERSOURCE,
+                                      fValue=2)  # 2 sets the trigger source to be software trigger.
+            assert abs(v - 2) < 0.00001
+        else:
+            raise ValueError('master pulse trigger is set to '+str(self.master_pulse_trigger)+';'
+                             'only \'CONTINUOUS\', \'EXTERNAL\' and \'SOFTWARE\' are supported.')
+
+        # allocate buffer
+        dcam.buf_alloc(self.buffer_size_frame_number - 1)
+
+    def get_ready(self, camera_ids=[0]):
+        """
+        This will checkout the camera with the specified camera_id.
+        With the Hamamtsu API, this would have to include the following steps:
+        1. initiate the dcamapi.
+
+        I'm implementing this with similar "pattern" as I used in Daxi-controller for now - Xiyu.
+
+        :return:
+        """
+        # 1. initiate the Dcamapi
+        Dcamapi.init()
+
+        for camera_id in camera_ids:
+            # 2. create the camera device with the camera index
+            self.devices = {'camera '+str(camera_id): Dcam(camera_id)}
+
+            # 3. open the camera device
+            self.devices['camera '+str(camera_id)].dev_open()
+
+            # 4. make sure the camera is opened
+            assert self.devices['camera '+str(camera_id)].is_opened()
+
+    def start(self, camera_ids=[0]):
+        """
```

Comment:
we will need to refactor this method and the following methods that shares similar purpose.